### PR TITLE
feat(tests): add initial integration tests to repo

### PR DESCRIPTION
**Pull request by ammirsm**
# 📝 Description

- Added new requirements.txt for the development, as we are not using poetry I just added this which has the requirements of the poetry for development and pytest running.
- Add new test folder for testing the integrations.
- Add new integration test which is basically replica of the intro.ipynb but inside the pytest to run it in CI
- Add integration tests to be ran in the CI.

## 🛠 Type of Change

- [x] New feature (non-breaking change which adds functionality)

## 🔍 How Has This Been Tested?

- [x] Added new unit/integration tests

## ✅ Suggested Checklist:

- [x] I have updated the `requirements.txt` file
- [x] New and existing unit tests pass locally with my changes
- [x] Pre-Commit checks are passing (locally and remotely)
- [x] Title of your PR / MR corresponds to the required format
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I ran `ruff check . --fix-only` to appease the lint gods
- [x] Commit message follows required format {label}(dspy): {message}

## ⚠️ Warnings

We need to clean up and move to poetry completely, some places we are using poetry and some places requirements which is making it very hard to maintain dependencies.

### CODE REVIEW

File: tests_integration/test_intro.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        # Call the predictor on a particular input
+        pred = generate_answer(question=dev_example.question)
+
+        # Assertions to verify the basic QA functionality
+        assert (
+            f"Question: {dev_example.question}"
+            == "Question: What is the nationality of the chef and restaurateur featured in Restaurant: Impossible?"
+        )
+        assert f"Predicted Answer: {pred.answer}" == "Predicted Answer: American"
+
```
Comment (arnavsinghvi11) [inline tests_integration/test_intro.py:216]: probably worth adding an assertion for the `turbo.inspect_history` behavior as well to ensure no changes are made to the LM interface (besides whatever is already present). I know we added some behavior in [this PR](https://github.com/stanfordnlp/dspy/pull/622) that's not reflected in the notebook but we should calibrate that here going forward

File: tests_integration/test_intro.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+            == "Restaurant: Impossible | Restaurant: Impossible is an American reality television series, featuring "
+            "chef and restaurateur Robert Irvine, that aired on Food Network from 2011 to 2016."
+        )
+        assert (
+            top_k_passages[1]
+            == "Jean Joho | Jean Joho is a French-American chef and restaurateur. He is chef/proprietor of Everest in "
+            "Chicago (founded in 1986), Paris Club Bistro & Bar and Studio Paris in Chicago, The Eiffel Tower "
+            "Restaurant in Las Vegas, and Brasserie JO in Boston."
+        )
+        assert top_k_passages[2][:30] == "List of Restaurant: Impossible"
```
Comment (arnavsinghvi11) [inline tests_integration/test_intro.py:193]: small nitpick but any reason not to check the full passage and retrieved value below? :)

File: tests_integration/test_intro.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        # Also check that the retrieved context actually contains that answer.
+        def validate_context_and_answer(example, pred, trace=None):  # noqa
+            answer_em = dspy.evaluate.answer_exact_match(example, pred)
+            answer_pm = dspy.evaluate.answer_passage_match(example, pred)
+            return answer_em and answer_pm
+
+        # Set up a basic teleprompter, which will compile our RAG program.
+        teleprompter = BootstrapFewShot(metric=validate_context_and_answer)
+        # Compile the RAG model
+        compiled_rag = teleprompter.compile(RAG(), trainset=trainset)
```
Comment (arnavsinghvi11) [inline tests_integration/test_intro.py:50]: is there a way we can check the stdout during compilation and assert the existing printed message? `Bootstrapped 4 full traces after 11 examples in round 0.`

File: tests_integration/test_intro.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+        assert f"Question: {my_question}" == "Question: What castle did David Gregory inherit?"
+        assert f"Predicted Answer: {pred.answer}" == "Predicted Answer: Kinnairdy Castle"
+        assert (
+            f"Retrieved Contexts (truncated): {[c[:10] + '...' for c in pred.context]}"
+            == "Retrieved Contexts (truncated): ['David Greg...', 'Gregory Ta...', 'David Greg...']"
+        )
+
+        # Verify compiled model's parameters
+        for name, parameter in compiled_rag.named_predictors():
+            assert name is not None
```
Comment (arnavsinghvi11) [inline tests_integration/test_intro.py:66]: i think it's good to also check that the `name` and `parameter.demos[0]` match up with the cached objects from the notebook.

File: .github/workflows/run_tests.yml [hunk 5]
```diff
@@ -120,3 +119,27 @@ jobs:
           cache: "poetry"
       - name: Run setup.py build
         run: python setup.py build
+
+  integration_test:
+    name: Run Integration Tests
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ["3.9"]
```
Comment (arnavsinghvi11) [inline .github/workflows/run_tests.yml:128]: not fully sure but is it needed to specify a python version here? (can we extend this to >=3.9 as in dspy-ai?) and how does this affect users who don't have 3.9 - (saw a followup to this [issue here](https://github.com/stanfordnlp/dspy/issues/802) more so for dspy-ai but similar impact).

**Comment by arnavsinghvi11** [comment]
Thanks @ammirsm ! this looks great - excited to have these integration tests on board! 
left a few general comments to add comprehensiveness. 

Regarding the poetry dependencies, is the plan to still support both `requirements.txt` and `poetry` dependencies? it was a bit unclear from the comments above. Also, will we remove the extra [`requirements.txt`](https://github.com/stanfordnlp/dspy/pull/1048/files#diff-2b4945591edfeaa4cf4d3f155e66d4b43d1bda7a55d881d5cf3107f1b05abbbc) before merging the PR?

Additionally, what's the best way to test the integration tests in this PR? Would we have to create mock PRs or run GitHub actions locally? would love to have that documented as well as this PR is being reviewed.

Also probably good to squash the commit history here :).

**Review by ammirsm — COMMENTED** [review]
> Thanks @ammirsm ! this looks great - excited to have these integration tests on board!
left a few general comments to add comprehensiveness.

thanks @arnavsinghvi11 for reviewing the PR. just replied to your comments and added some changes to the PR regarding to them.

> Regarding the poetry dependencies, is the plan to still support both requirements.txt and poetry dependencies? it was a bit unclear from the comments above. Also, will we remove the extra [requirements.txt](https://github.com/stanfordnlp/dspy/pull/1048/files#diff-2b4945591edfeaa4cf4d3f155e66d4b43d1bda7a55d881d5cf3107f1b05abbbc) before merging the PR?

In regard to poetry dependencies, we plan to eventually move everything to poetry as it's faster in the CI. After this PR is merged and auto deployment is integrated (https://github.com/stanfordnlp/dspy/pull/919), we'll proceed with the transition to poetry, which has been initiated here: https://github.com/stanfordnlp/dspy/pull/819

To clarify further:
Currently, we don't have any way to develop using pytest with our original requirements.txt. All dependencies for development and document generation are in poetry. I've added a new requirements-dev.txt to include all existing requirements for development there. This allows us to maintain everything in CI using pip installer for now. Once we've stabilized the requirements, we'll transition everything to poetry.

> Additionally, what's the best way to test the integration tests in this PR? Would we have to create mock PRs or run GitHub actions locally? would love to have that documented as well as this PR is being reviewed.

I didn't quite understand this one but will try to add some clarification, lmk if I haven't got your question right. 
To run the tests, you simply execute 'pytest tests_integration' locally. When do these tests run? They're executed with each merge and pull request to ensure they don't break our package. Additionally, Hanna will add another workflow that tests the package installed from PyPi, ensuring our package on PyPi is also functioning correctly.

> Also probably good to squash the commit history here :).

haha! definitely will squash it for the merge.

File: tests_integration/test_intro.py (hunk 2 above)
Comment (ammirsm) [reply tests_integration/test_intro.py]: anyways changed it to full paragraph 🗡️

**Comment by arnavsinghvi11** [comment]
Thanks @ammirsm  - the comments look good from my end!

>added a new requirements-dev.txt to include all existing requirements for development there. This allows us to maintain everything in CI using pip installer for now. Once we've stabilized the requirements, we'll transition everything to poetry.

I see so just to clarify, we have a temporary `requirements-dev.txt` for now that will help with the impending poetry migration, and once this is done, we will remove that to avoid the duplicate `requirements.txt` files?

> To run the tests, you simply execute 'pytest tests_integration' locally. When do these tests run? They're executed with each merge and pull request to ensure they don't break our package. Additionally, Hanna will add another workflow that tests the package installed from PyPi, ensuring our package on PyPi is also functioning correctly.

yup this is perfect!

**Comment by ammirsm** [comment]
thanks for the reply @arnavsinghvi11 .

> I see so just to clarify, we have a temporary `requirements-dev.txt` for now that will help with the impending poetry migration, and once this is done, we will remove that to avoid the duplicate `requirements.txt` files?

yes sir, basically bc we don't want to install dev dependencies during the package installation we put them in the new requirements file and we will remove it when we moved to poetry.

**Pull request merged**